### PR TITLE
BUG: Fix query doesn't support equals or not equals with Python parse…

### LINKED ISSUE #40436: BUG: DataFrame.query doesn't support equals or not equals for string columns with python parser
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

---

#### Code Sample, a copy-pastable example

```python
In [3]: import pandas as pd

In [4]: pd.__version__
Out[4]: '1.2.3'

In [5]: import numpy as np

In [7]: x = np.random.choice(["foo", "bar", "baz"], size=10, replace=True)

In [8]: df = pd.DataFrame({"x": x})

In [9]: df
Out[9]: 
     x
0  bar
1  baz
2  foo
3  baz
4  bar
5  baz
6  baz
7  bar
8  foo
9  foo

In [10]: df.query("x == 'bar'")
Out[10]: 
     x
0  bar
4  bar
7  bar

In [11]: df.query("x == 'bar'", parser="python")
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)

...

~/malariagen/binder/conda/envs/xarray-tests/lib/python3.8/site-packages/pandas/core/computation/expr.py in f(self, *args, **kwargs)
    251 
    252     def f(self, *args, **kwargs):
--> 253         raise NotImplementedError(f"'{node_name}' nodes are not implemented")
    254 
    255     return f

NotImplementedError: 'In' nodes are not implemented

In [12]: df.query("x != 'foo'")
Out[12]: 
     x
0  bar
1  baz
3  baz
4  bar
5  baz
6  baz
7  bar

In [13]: df.query("x != 'foo'", parser="python")
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)

...

~/malariagen/binder/conda/envs/xarray-tests/lib/python3.8/site-packages/pandas/core/computation/expr.py in f(self, *args, **kwargs)
    251 
    252     def f(self, *args, **kwargs):
--> 253         raise NotImplementedError(f"'{node_name}' nodes are not implemented")
    254 
    255     return f

NotImplementedError: 'NotIn' nodes are not implemented
```

#### Problem description

If the query expression has an equals or not equals comparison with a string, and the "python" parser is selected, then a `NotImplemented` error is raised. The error messages however reports that "In" nodes are not implemented, or "NotIn" nodes are not implemented, which doesn't fit with the expression, suggesting a bug in how the query expression is being parsed.

#### Expected Output

Expected output should match what you get with the "pandas" parser as semantics for this type of query expression should be the same for either parser.

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit           : f2c8480af2f25efdbd803218b9d87980f416563e
python           : 3.8.8.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.4.0-66-generic
Version          : #74~18.04.2-Ubuntu SMP Fri Feb 5 11:17:31 UTC 2021
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_GB.UTF-8
LOCALE           : en_GB.UTF-8

pandas           : 1.2.3
numpy            : 1.20.1
pytz             : 2021.1
dateutil         : 2.8.1
pip              : 20.2.4
setuptools       : 49.6.0.post20210108
Cython           : None
pytest           : 6.2.2
hypothesis       : 6.8.0
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.6.2
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.3
IPython          : 7.21.0
pandas_datareader: None
bs4              : 4.9.3
bottleneck       : 1.3.2
fsspec           : 0.8.7
fastparquet      : None
gcsfs            : None
matplotlib       : 3.3.4
numexpr          : 2.7.3
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : 1.6.0
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : 0.17.1.dev3+g48378c4b
xlrd             : None
xlwt             : None
numba            : 0.52.0
</details>

### PR DESCRIPTION
**Pull request by alexprincel**
…r (#40436)

- [X] closes #40436 
- [X] tests added / passed
- [X] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [X] whatsnew entry

### CODE REVIEW

**Comment by alexprincel** [comment]
Could some of the pandas maintainers let me know if the failed checks are normal ? They are unrelated to any changes I've made as far as I know.

**Comment by mzeitlin11** [comment]
> Could some of the pandas maintainers let me know if the failed checks are normal ? They are unrelated to any changes I've made as far as I know.

If you look at the failures in a run like https://github.com/pandas-dev/pandas/pull/40563/checks?check_run_id=2163292906, looks like they might be related?

**Comment by jbrockmendel** [comment]
They look related to me.  Do these tests pass for you locally?

File: pandas/tests/frame/test_query_eval.py [hunk 1]
```diff
@@ -548,6 +548,13 @@ def test_query(self):
             df[df.a + df.b > df.b * df.c],
         )
 
+    def test_query_equality(self):
+        engine, parser = self.engine, self.parser
+        df = DataFrame({"x": np.random.choice(["A", "B", "C"], size=20)})
+        tm.assert_frame_equal(
+            df.query("x == 'A'", engine=engine, parser=parser), df[df.x == "A"]
```
Comment (jreback) [inline pandas/tests/frame/test_query_eval.py:555]: don't write compound asserts, use 
result=
expected=
tm.assert_frame_equal(result, expected)

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by alexprincel** [comment]
> > Could some of the pandas maintainers let me know if the failed checks are normal ? They are unrelated to any changes I've made as far as I know.
> 
> If you look at the failures in a run like https://github.com/pandas-dev/pandas/pull/40563/checks?check_run_id=2163292906, looks like they might be related?

I understand that my changes had a broader impact than I had anticipated. Some of these tests hard coded the == and != operators as NotImplemented, which I have undone and this is causing them to fail. 

Will have to really dig around to understand if these changes I made are ok but simply failing tests which should be amended or if I may be introducing some bug(s). Will take much more time than anticipated.

**Comment by alexprincel** [comment]
Ok, I dug further into this and I both understand more and less about the issue. I now understand that both the `pandas` and `python` visitors inherit from the same `BaseExprVisitor` albeit with different parameters for `disallow` and different `preparser` function.

However, I do not understand why `In` and `NotIn` were specifically disallowed for the `python` engine even though its treatment would be identical under both `PythonExprVisitor` and `PandasExprVisitor`. This appears quite intentional as removing `In` and `NotIn` from the `PythonExprVisitor` disallow list makes a few tests fail as they specifically expect queries using these bin ops to raise `NotImplemented`.

Could someone please educate me as to why this is desirable ?

**Comment by jreback** [comment]
would take this if you can merge master and will look again

**Comment by mroeschke** [comment]
Thanks for the PR, but it appears to have gone stale. If interested in continuing, please merge master and a maintainer can have a second look.

**Pull request closed without merging**